Added support for outlined text

zsbzsb opened the pull request:
Considering all the requests [over the years](https://www.google.com/search?q=sfml+text+outline+site:sfml-dev.org) for outlined text ([and the totally hackish workarounds](http://en.sfml-dev.org/forums/index.php?topic=15071.msg106581#msg106581)) and also that my current project will definitely need this at some point, I went ahead and implemented it. I made the PR for code review.

Some key points are the following.

- Outlined glyphs are stored in the same page as normal glyphs (can be changed if this is deemed that it will fill up the page texture too fast).
- ```Text::[set|get]Color(...)``` is now marked as deprecated, instead everyone should use ```Text::[set|get]FillColor(...)```
- ```Font::getGlyph(...)``` has a new parameter, outline thickness which is defaulted to 0 (to avoid API breakage). The parameter determines how thick the outline will be in pixels.
- Since outlined glyphs are stored in the same page/texture the key type has been changed to Int64 to continue being unique now that the outline thickness is part of the key.
- I refactored ```Text::ensureGeometryUpdate()``` and split the duplicated code to append vertices to separate functions ```::addLine(...)``` and ```::addGlyphQuad(...)```. These can be renamed if someone could think of more appropriate names.
- ~~Outline thickness is unsigned, I don't see any easy ways with freetype to make the outline protrude inwards with a negative thickness.~~ It is a float, but a negative value causes distorted rendering.
- Nothing should change with the current rendering of text if the user decides not to use outlines.

As stated before, please review this and give some feedback.

```
#include <SFML/Graphics.hpp>

int main()
{
    sf::RenderWindow window({ 700, 400 }, "Outline test");
    window.setVerticalSyncEnabled(true);

    sf::Font font;
    font.loadFromFile("C:/Windows/Fonts/arial.ttf");

    sf::Text text(L"\u0476ZyG\noLPfkdf", font, 140);
    text.setPosition({ 20, 20 });

    text.setFillColor(sf::Color::Red);

    text.setOutlineThickness(5);
    text.setOutlineColor(sf::Color::White);

    text.setStyle(sf::Text::Underlined | sf::Text::Italic | sf::Text::StrikeThrough);

    while (window.isOpen())
    {
        sf::Event e;
        while (window.pollEvent(e))
        {
            if (e.type == sf::Event::Closed)
                window.close();
        }

        window.clear();

        window.draw(text);

        window.display();
    }
}
```

![Example Image](http://i.imgur.com/crqgL0x.png)

MarioLiebisch commented:
Sounds interesting. :) Just skipped through the code. You're essentially creating two glyphs, one for the inside, one for the outline?

Out of interest (haven't had time to try it): Is the outline really just the outline? Or does it include parts covered by the inner part of the font. In essence, could you render the outline only while keeping the inside transparent?

zsbzsb commented:
>You're essentially creating two glyphs, one for the inside, one for the outline? Yes, its the only way to colorize the outline differently.

>Is the outline really just the outline? Or does it include parts covered by the inner part of the font. In essence, could you render the outline only while keeping the inside transparent?

I was wondering the same thing, actually freetype when stroking the outline draws on both sides of the character border/glyph. So if you don't render the fill (or set to transparent) the outline will extend to both sides (up to outline thickness) of the character's outline. I don't see any way to make it so that it doesn't do that.

I also just added a note in the documentation about this.

MarioLiebisch commented:
Ah, well, that's partially what I wanted. :)

But yeah, no issue, just been curious about this.

zsbzsb commented:
@LaurentGomila @eXpl0it3r @binary1248 @MarioLiebisch @mantognini @TankOs 

If no one has anything against this change, can we try for a 2.3 inclusion?

MarioLiebisch commented:
Haven't had a chance to try it yet, but I definitely like it (and planned to do something similar in the past).

LaurentGomila commented:
I have nothing to say against this PR.

mantognini commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -299,14 +375,45 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
+    ///
+    /// This used for adding strikethrough and underline lines
+    ///
+    /// \param position         Width and top of the line
```
"Left and top", no?

zsbzsb replied:
The documentation is correct. It is the width and top right now as it is implemented and there is a 3rd parameter for the vertical offset. However looking over this is does make sense to change it to "left and top" and change the 3rd parameter to the width of the line... unless I am forgetting something from the last time I looked at this.

Bromeon replied:
"Position" sounds a bit misleading if you're actually referring to a (width, top) coordinate pair. Is there no better term?

TankOs replied:
*bump* :grinning:

zsbzsb replied:
Sorry for the delay (had to re-familiarize myself with the code)... the reason it was like this is because in the ```Text::ensureGeometryUpdate()``` function the current character position is tracked as ```x``` and ```y``` variables. It also happened to be that ```x``` was also the width of current text line when adding an outline/underline line (need more "line"s?) and ```y``` was the top.

So when I wrote the ```addLine(...)``` function I grouped the two variables (now as width and top) together in a vector. Looking over the code it seems now more reasonable to split that vector and name the variables ```lineLength``` and ```lineTop```. The PR now reflects these changes.

mantognini commented:
Tested quickly on Ubuntu, seems to run smoothly. :+1: 

Beside my to comments, the code in `sf::Font` could use some refactoring but it's not really related to this PR so... Commits need to be squashed too. Then we could include it in 2.4 I guess (can someone @SFML/sfml validate this and set the milestone?).

zsbzsb commented:
```Commits need to be squashed too```

Can be done once the changes are verified.

```Then we could include it in 2.4 I guess (can someone @SFML/sfml validate this and set the milestone?).```

Sounds great :+1:

Bromeon commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -159,18 +159,61 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
+    /// \see getOutlineThickness
+    ///
+    ////////////////////////////////////////////////////////////
+    void setOutlineThickness(unsigned int outlineThickness);
```
Is there a technical limitation that allows only integral thicknesses?
I'm asking because this API is inconsistent with `Shape::setOutlineThickness(float thickness)`.

(minor thing: I would at least call the parameter the same, i.e. `thickness` instead of `outlineThickness`)

zsbzsb replied:
>Is there a technical limitation that allows only integral thicknesses?

[```FT_Stroker_Set(...)```](http://www.freetype.org/freetype2/docs/reference/ft2-glyph_stroker.html#FT_Stroker_Set) takes a [```FT_Fixed```](http://www.freetype.org/freetype2/docs/reference/ft2-basic_types.html#FT_Fixed) as the radius which is typedef(fed) as signed long. However I restricted the thickness to unsigned as passing a negative value does not look like you would expect.

>minor thing: I would at least call the parameter the same, i.e. thickness instead of outlineThickness

Sure, easily changed :smile:

Bromeon commented on src/SFML/Graphics/Font.cpp:
```diff
@@ -143,6 +144,15 @@ bool Font::loadFromFile(const std::string& filename)
 
+    // Load the stroker that will be used to outline the font
+    FT_Stroker stroker;
+    if (FT_Stroker_New(static_cast<FT_Library>(m_library), &stroker) != 0)
```
The stroker resource is not released if one of the subsequent allocations (e.g. charmap) fails.

zsbzsb replied:
```cleanup``` is always called in the destructor and ```loadFromFile(...)```. I didn't change any of the program flow regarding the cleanup function so if there is a problem with it then it has been in the code base for quite some time. ;)

TankOs commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -159,18 +159,61 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
+    ////////////////////////////////////////////////////////////
+    /// \brief Set the outline color of the text
+    ///
+    /// By default, the text's color is opaque black.
```
By default, the text's **outline** color is opaque black.

zsbzsb replied:
Fixed

zsbzsb commented:
Anything else? Is the ```lineLength``` and ```lineTop``` naming fine now for the ```addLine(...)``` function?

TankOs commented:
:+1: :cat2: 
Seems good to me, well done.

Foaly commented on src/SFML/Graphics/Text.cpp:
```diff
@@ -231,6 +294,11 @@ void Text::draw(RenderTarget& target, RenderStates states) const
         states.texture = &m_font->getTexture(m_characterSize);
+
+        // Only draw the outline if there is something to draw
+        if (m_outlineThickness > 0)
```
I think this can be shortend to `if (m_outlineThickness)` because `m_outlineThickness` is an `unsigned int`.

LaurentGomila replied:
We don't use this kind of shortcuts, it just makes the code less readable for no benefit.

Foaly replied:
Ok. I totales agree with you. But if (m_outlineThickness) is used everywhere else in the code to see if the outline is used or not. I just wanted it to be consistent. But I agree if (m_outlineThickness > 0) is more expressive.

MarioLiebisch replied:
Well, in this case I consider both telling you essentially the same thing: "If we've got an outline thickness, then...".

Bromeon replied:
> But if (m_outlineThickness) is used everywhere else in the code to see if the outline is used or not.

Then that's a mistake that must be corrected. We only use "implicit" checks for `bool` and pointers, see [code style](http://www.sfml-dev.org/style.php#type-conversions).

Foaly replied:
Like I said I only wanted to point out, that there is an inconsistency. But I totally agree `if (m_outlineThickness > 0)` is more expressive.

zsbzsb replied:
For reference, [```sf::Shape```](https://github.com/SFML/SFML/blob/master/src/SFML/Graphics/Shape.cpp#L222) contains the only other check for outline thickness and it uses a ```!=``` comparison. I can change the implementation here to match that if I understand correctly that nobody wants to see ```if (m_outlineThickness)```.

zsbzsb replied:
I updated all ifs that involved the outline thickness to use a ```!= 0``` check.

zsbzsb commented:
Any chance we could get this merged so @Foaly could get #928 updated?

LaurentGomila commented:
Looks good.

It feels weird to have an integer outline thickness, but there's no other way since it's used in the lookup key.

zsbzsb commented:
>It feels weird to have an integer outline thickness

Blame freetype's stroker API for that.

>there's no other way since it's used in the lookup key

That wasn't really an issue, if we had to use a float as part of the key we could have fun with casting ([same way ```sf::Packet``` handles things](https://github.com/SFML/SFML/blob/master/src/SFML/Network/Packet.cpp#L243)).

Bromeon commented:
Why reinterpret-casting? Wouldn't static-casting suffice?

It's of course an annoying limitation, but should we expose the implementation details? What if we switch to another backend in the future (if there is one) that supports floats? :smiley:

LaurentGomila commented:
> Why reinterpret-casting? Wouldn't rounding and static-casting suffice?

We would lose the benefit of floats then, because the range [x - 0.5 ... x + 0.5] of thicnkess will all map to the same set of glyphs (x). It's used as a lookup key, so we really need one entry per different outline thickness value.

> It's of course an annoying limitation, but should we expose the implementation details?

Since implementation doesn't allow us to implement something else... yes :p

MarioLiebisch commented:
How about accepting a `float` parameter, but documenting that it only supports like `0.1` steps (i.e. multiply by 10)? This limitation could later on be lifted.

Or we could just accept `float` and say it actually only supports integer values? This sounds like the worse idea to me though.

eXpl0it3r commented:
I'd rather go with an integer and if we'd ever change the implementation (very unlikely) we could mark it as deprecated.

LaurentGomila commented:
> How about accepting a float parameter, but documenting that it only supports like 0.1 steps (i.e. multiply by 10)?

And how do you support 0.1 steps when the implementation only accepts integers?

> I'd rather go with an integer and if we'd ever change the implementation (very unlikely) we could mark it as deprecated.

:+1:
It's even simpler, since switching from int to float doesn't break API compatibility.

Bromeon commented:
Okay. I mentioned this with respect to API consistency between `sf::Shape` and `sf::Text` (the former has `void setOutlineThickness(float)`. It's impossible to use a virtual function in the future, should we ever see that an abstraction within SFML makes sense.

It makes user-side abstractions more difficult too, but at least we give user the control and responsibility to convert floats into ints.

MarioLiebisch commented:
> And how do you support 0.1 steps when the implementation only accepts integers?

    int index = static_cast<int>(float * 10);

As far as I know the only part limited to integers is the index for the glyphs?

LaurentGomila commented:
> As far as I know the only part limited to integers is the index for the glyphs?

Noooo :grin:
As @zsbzsb said: *Blame freetype's stroker API for that.*

But... after checking, ```FT_Stroker_Set``` takes a fixed-point decimal number for its ```radius``` parameter. So where exactly is the integer limitation?

zsbzsb commented:
```FT_Stroker_Set``` takes a [```FT_Fixed```](http://www.freetype.org/freetype2/docs/reference/ft2-basic_types.html#FT_Fixed) which is typedef'd to ```typedef signed long``` so how exactly do you plan on passing a float to that? Then there is the problem that negative values don't render how the user would expect them to render.

LaurentGomila commented:
FreeType uses fixed-point (as opposed to floating point) numbers. It's just a different way to encode real numbers.

I couldn't find how to convert a float to fixed-point in their doc, but there should be something (maybe even a macro).

We already do the other way round (fixed to floating) in Font.cpp, for example:
```cpp
static_cast<float>(face->size->metrics.height) / static_cast<float>(1 << 6);
```

zsbzsb commented:
Well thanks to that little heads up information, I was able to change the implementation to support float outlines. Now there is still the issue that users can specify a negative value and they won't get the expected result... should I add a note about that in the docs?

zsbzsb commented:
>Assuming all our target platforms have 32-bits floats, this would rather be:

Wouldn't we need to do a static cast to a Uint64 then before bit shifting?

>Yes. (what do negative values do?)

![Image](http://i.imgur.com/CImn3Fm.png)

LaurentGomila commented:
> Wouldn't we need to do a static cast to a Uint64 then before bit shifting?

Probably. I don't remember the rules in this case.

> &lt;ugly stuff&gt;

Ok :stuck_out_tongue:

eXpl0it3r commented:
So is this ready or do we need to change something?

Foaly commented:
I think we should prohibit negative outline thickness, since it doesn't make sense and returns ugly results. So if negative values are passed set it to 0. 
Other than that everything looks good!

Bromeon commented:
> I think we should prohibit negative outline thickness, since it doesn't make sense and returns ugly results. 

Agreed.

> So if negative values are passed set it to 0. 

**No. Never do that, it just hides logic errors.** Use an assertion (at least for the moment, until we have decided on a consistent way to handle such errors).

zsbzsb commented:
I have now added a note to the docs about using a negative outline thickness (as float outlines are supported). I have also avoided introducing any error checking as @Bromeon suggested.

@eXpl0it3r this should be ready as all issues have been addressed. Unless someone wants to find something else to change :)

Bromeon commented:
I created a little editor to experiment with outlined texts here: https://gist.github.com/Bromeon/d9510ecfd0b6884ce88f

The functionality works well, apart from two minor issues:
* Wenn you change the outline thickness with F5/F6, you'll see that it doesn't appear continuous. The outline seems to shrink sometimes when it should grow. I'm not sure if this is a problem of OpenGL rasterization, or if there is an error with rounding or so.
* When a newline is appended at the end, and strikethrough or underline style is used, then the outline of that `\n` whitespace is drawn on the next line (even though there's no text to outline).

zsbzsb commented:
@Bromeon Thanks for testing and writing a small test to find these issues :smile:

>Bold text is not outlined correctly. This would need a special case, because of the way in which bold text is implemented.

This was actually another issue with the glyph key creation. For some reason the bitshifting wasn't working right when shifting 32bit integers and then combining into a 64 bit key. So I wrapped each part in a ```static_cast<Uint64>(...)``` and now it works correctly.

>When** you change the outline thickness with F5/F6, you'll see that it doesn't appear continuous. The outline seems to shrink sometimes when it should grow. I'm not sure if this is a problem of OpenGL rasterization, or if there is an error with rounding or so.

Looking at this it seems to be a mix of FreeType stroking/rasterization and floating point inaccuracy. Rounding the outline coordinates (when rendering) only seemed to make the problem worse so I don't think it is OpenGL rasterization. If you look closely you will notice that the outline does enlarge/shrink as requested, the issue is that it moves just slightly around the stationary text fill which gives the illusion that it is shrinking when it should be growing (and vice versa).

I also tried rendering at 2x the size and then shrinking the rendered text and it is hardly an issue. This alone makes me think it comes down to floating point inaccuracy and is something we can't exactly fix. Let's also be realistic, most users are going to be using such small values changes for the outline thickness. If you take your test code and increment the outline thickness by .4f the issue almost entirely disappears.

>When a newline is appended at the end, and strikethrough or underline style is used, then the outline of that ```\n``` whitespace is drawn on the next line (even though there's no text to outline).

Fixed, this was a fault of the original code flow that it would actually append an invisible line on the ending blank line. I corrected this by adding a check to make sure the line has a length before adding any lines. It may even make text rendering slightly faster.

eXpl0it3r commented:
And last reviewing/testing round? :wink:

Foaly commented:
I did a test with the latest version of this branch and the test code that Bromeon provided.
Everything looks good and seems to work fine. 

The only thing I noticed was that starting at a outline thickness of 16 you start to see weird artefacts around dots (like the dot on the i or with a colon). But I guess that is the stroker from FT and we can do much about that. Example image:
![unbenannt](https://cloud.githubusercontent.com/assets/1014762/10178286/253cd724-66ff-11e5-8df5-b205ef9a4f4c.png)

Bromeon commented:
I also experience the artifact mentioned by Foaly at several punctuation characters. It only appears at extreme outlines though.

Other than that, it seems good -- bold text and newlines at the end have been fixed!

LaurentGomila commented:
Do we wait for the deprecation macro PR to be merged, or do we merge this one now and add the deprecated stuff in a new PR?

zsbzsb commented:
I can also reproduce the artifacts and looking and the glyph texture can confirm it is how FreeType is stroking the glyph (they don't seem to handle intersecting vertices very well).

Bromeon commented:
> Do we wait for the deprecation macro PR to be merged, or do we merge this one now and add the deprecated stuff in a new PR?

I wouldn't mind if https://github.com/SFML/SFML/pull/969 were merged soon, it's something that's required by various upcoming features. Also, the deprecation is a really small and not API-concerning modification, so it requires less review and discussion than actual features like this PR.

> I can also reproduce the artifacts 

Ok. I think this pull request shouldn't be held back because of minor imperfections like that. If it really matters, we can still fix/workaround it at a later stage. But it's well possible that no one even notices in real-world scenarios.

If somebody else could review the code, that would be nice!

binary1248 commented:
I don't know about you guys, but I prefer having `Text::addLine()` and `Text::addGlyphQuad()` defined as free functions local to the translation unit. They don't have to access any members and are just adding to the external interface (even though they are private).

Bromeon commented:
I agree with @binary1248, an anonymous namespace would be ideal. Furthermore, those functions should take the container as the first parameter.

zsbzsb commented:
So if I understand correctly, you want to see ```Text::addLine(...)``` and ```Text::addGlyphQuad(...)``` moved to an anonymous namespace and also the vertex array as the first argument?

Bromeon commented:
Exactly :smiley:

eXpl0it3r commented:
Any progress, @zsbzsb?

zsbzsb commented:
I was waiting for the deprecation macro to be merged, so I will be updating this in the next few days.

eXpl0it3r commented:
What's the upper limit of days for "few"? :stuck_out_tongue_winking_eye:

zsbzsb commented:
@eXpl0it3r All changes have now been made. ```Text::[get|set]Color(...)``` has been marked as deprecated with the new macro and the examples were also updated. Also the ```addLine(...)``` and ```addGlyphQuad(...)``` functions have now been moved into an anonymous namespace.

zsbzsb commented:
@SFML Anything else?

Bromeon commented on include/SFML/Graphics/Text.hpp:
```diff
@@ -224,14 +270,47 @@ class SFML_GRAPHICS_API Text : public Drawable, public Transformable
+    /// \see setFillColor
+    ///
+    /// \deprecated This function is deprecated and may be removed in future releases.
+    /// Use getFillColor instead.
```
Minor remark: the line "This function is deprecated and may be removed in future releases." does not add any information to the `\deprecated` command, I'd rather say something about the backgrounds, i.e. that there are now 2 different colors to set -- fill and outline.

Also, I'd use `getFillColor()` with parentheses at the end for consistency (and I'm not sure if auto-link generation works otherwise).

zsbzsb replied:
I changed it to "There is now fill and outline colors instead of a single global color."

Bromeon commented:
It looks mostly good. I haven't tested the code again now -- there have not been any functional changes since the last iteration, right?

eXpl0it3r commented:
This PR has been added to my merge list, meaning it will be merged soon, unless someone raises any concerns.

zsbzsb commented:
All is done..... again :grinning:

eXpl0it3r commented:
Merge in 957cabb816f69a3d7a3909d24b9e773ffc9a65b5

hobby8 commented:
Indeed some parts will need to be rewritten due to #840 - should be quite easy to do.
